### ISSUE TITLE
Standardise date formatting

### ISSUE THREAD

**Issue by mtthwhggns** [apluslms/a-plus#572]
We currently appear to change date format by language, which is, in my opinion, incorrect. Date formatting is region-specific, not language-specific. 

In my opinion, a Finnish University (for example), should write dates `11.06.2020` whether writing in English, Finnish or Swedish. It definitely doesn't make sense to use the American format (06/11/2020), as this is a huge source of confusion. 

Alternatively, given our target users' backgrounds, maybe ISO-8601 would be even clearer... `2020-06-11`

>This PR may be too specific for this issue, but in my opinion, we should not be changing the date format by language, nor should we be using the American date format at all. 
>
>Would you like to look at that, or create a new issue somewhere else?

_Originally posted by @mtthwhggns in https://github.com/apluslms/a-plus/pull/560_

**Comment by Mankro** [comment]
Is the fix for this issue going to change mostly frontend HTML templates or some backend configuration? I was first thinking of backend, so I assigned this small task to @veeraalt.

**Comment by Mankro** [comment]
We don't want to hardcode any specific date format. So, date format should be based on the region. Does Django have standard settings for defining the region or locale for dates independent of the language used in the UI?

**Comment by raphendyr** [comment]
Should be doable with the Django's format system. Basically, it allows defining locale settings per language. I'm not sure if it would be possible to have two different english languages with different formats, if that would ever be needed.

The documentation : https://docs.djangoproject.com/en/2.2/ref/settings/#std:setting-FORMAT_MODULE_PATH

However, just disabling `USE_L10N` might be enough.. https://docs.djangoproject.com/en/2.2/ref/settings/#std:setting-USE_L10N

**Comment by veeraalt** [comment]
So do we want the date format to be based on the region or to use the same format for all users?

And if we don't want to hardcode any specific date format, should [aplus/en/formats.py](https://github.com/apluslms/a-plus/blob/master/aplus/en/formats.py) and [aplus/fi/formats.py](https://github.com/apluslms/a-plus/blob/master/aplus/fi/formats.py) be deleted (or left empty)?

**Comment by mtthwhggns** [comment]
In my opinion, each running instance of the software should have one consistent one. Maybe it should be configurable the same way that the site "brand" name is. 

They definitely shouldn't be different based on language or where we think the user is.

**Comment by Mankro** [comment]
I am seeing some conflicting expectations here. On one hand, I agree with Matthew. In Finland, we should use the Finnish date formatting and similarly for currency. Then the format does not change even if the content is displayed in English. I now noticed myself that indeed, the A+ front page changes the date format based on the language (that the user can set in the profile page). It is probably activated by the `USE_L10N` Django setting since Django supports automatic format localization (https://docs.djangoproject.com/en/2.2/topics/i18n/formatting/). It looks like that `DATE_FORMAT` (https://docs.djangoproject.com/en/2.2/ref/settings/#date-format) can be used when `USE_L10N = False`.

On the other hand, we may offer courses to people outside Finland and forcing them to see Finnish date formatting is not right. In that case, we would want to change formats based on the language of the content or the language chosen by the user. If we had user-specific settings for the preferred localization in the user profile, then the users could set their own locale. However, let's not spend time on this now.

To summarize, we can keep this simple and use settings in `settings.py` for defining the date format used in the site. The server admin can change the localization in the settings, so we do not hardcode the value in the source code, but the server can only use one localization at a time for all users. Our production A+ site will use the Finnish localization, but other servers could easily use another localization.

**Comment by veeraalt** [comment]
@Mankro Thanks for clearing things up!

>To summarize, we can keep this simple and use settings in settings.py for defining the date format used in the site. 

This sounds good.

**Comment by ttsirkia** [comment]
Instead of disabling the whole formatting of dates in A+ as part of #604 , just add `SHORT_DATE_FORMAT = 'M j Y'` for the English localization file. That will fix the front page. All other pages uses the nicely formatted strings containg the weekday etc.

**Comment by Mankro** [comment]
@ttsirkia does it require to keep `USE_L10N = True`? We wanted to keep the date formatting consistent despite the language of the content. Or are you actually saying that we would set identical date formats for Finnish and English? When the date includes a written weekday (Mon, Tue), it must be given in the same language as the content of the page,

**Comment by ttsirkia** [comment]
Yes it requires, but it doesn't make any sense to show the course contents in Finnish and weekdays in English or vice versa. I totally disagree this approach of making them look the same.

**Comment by mtthwhggns** [comment]
> Yes it requires, but it doesn't make any sense to show the course contents in Finnish and weekdays in English or vice versa. I totally disagree this approach of making them look the same.

To be clear, the "making them look the same" was with reference to the short date format only. There are two incompatible English versions in widespread use, `DD/MM/YY` and `MM/DD/YY`. These aren't tied to a language, (English), but to a location (UK or US mostly). 

The idea was that if we went with the formatting to match the location of the institution (so in Finland we have `DD.MM.YYYY`), we have consistency.

I agree about not showing weekdays (or month names) if written in full in English.

Most operating systems seem to split language and region. I use macOS in English with region set to Finland. That means I get numbers like `3,14` and `1 000 000`, and dates like `31.02.2020`. In my opinion we should have similar functionality here.

**Comment by ttsirkia** [comment]
If the only concern is the short date format, then it is possible to change that only to be the same in Finnish and English by adding the Finnish short format to the English version of settings. In that way it doesn't affect the weekday names and their language in any way. As far as I know, the front page is the only page which uses the short format. That could be easily `grep`ed.

**Comment by ttsirkia** [comment]
Disabling the `USE_L10N = True` would be a huge drawback.

**Comment by ttsirkia** [comment]
Using the short format in the front page is enforced here: https://github.com/apluslms/a-plus/blob/master/course/templates/course/_course_cards.html#L32

**Comment by Mankro** [comment]
Thanks @ttsirkia for the comments. 

@veeraalt could you try out different alternatives involving `USE_L10N` and some other format settings and write a short comparison of them? Which approach provides consistent date formatting and does not break the language of weekdays (e.g., if a date includes "Monday" in a Finnish text)? Even if `USE_L10N = True`, it should be possible to set the English and Finnish date formats to be similar. Is there a strong reason for keeping `USE_L10N = True` and what would we lose with the False value?

**Comment by ttsirkia** [comment]
I guess the biggest reason to use `USE_L10N = True` is to have the weekday and month names written in the correct language.

**Comment by Mankro** [comment]
I see, but is it really so that Django can not translate the weekday and month names without `USE_L10N = True`? I thought that the formats including weekdays would be filled in with the weekday name in the currently activated language. I have not tried this myself, but it would be good to verify this. Veera could try it out, please?

**Comment by raphendyr** [comment]
So, we have found out that we have different formats in different places. Some of those, e.g. Finnish long datetime format, are really strange. Thus, I think we need to do the following:

1. find all locations where we render dates,. times and datetimes and note the formatting and if it's good for the purpose.
2. with that analysis in mind, we can plan how to fix different locations. Maybe the Django date formats are fine, so let's use those where possible. If we need more formats, then we need to rethink.
3. find a way to make it possible to edit region based information per installation and do that easily
4. fix all the stupid formats and make them feel coherent
5. consider if we should wrap all dates to [time](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/time) element (need to check how IE 11 ignores them)

These steps can be handled with multiple PRs.

**Comment by raphendyr** [comment]
To my knowledge, `USE_L10N = True` should enable picking settings based on the active language, but the "human readable day" should be rendered in the active language even without it. Though, many date strings have a bit different ordering in different languages even in the same region.

**Comment by veeraalt** [comment]
> I see, but is it really so that Django can not translate the weekday and month names without USE_L10N = True?

I tried it out and the weekday and month names seem to be translated (when the course is available in the activated language) also when USE_L10N = False.

**Comment by veeraalt** [comment]
Here's a not so short comparison of different alternatives.

Date formats being used now:
- SHORT_DATE_FORMAT (English: `07/01/2020` and Finnish: `01.07.2020`) used in
	- front page, course opening time
	https://github.com/apluslms/a-plus/blob/master/course/templates/course/_course_cards.html#L32

- longer DATETIME_FORMAT (English: `Wed, Jul 1 2020, noon` and Finnish: `1.7.2020 keskiviikko 12:00`) used in
	- course module opening time
	https://github.com/apluslms/a-plus/blob/master/course/templates/course/module.html#L27
	- exercise deadlines
	https://github.com/apluslms/a-plus/blob/master/exercise/exercise_models.py#L447
	https://github.com/apluslms/a-plus/blob/master/exercise/templates/exercise/_user_results.html#L62
- 2020-07-01 12:00
  https://github.com/apluslms/a-plus/blob/master/exercise/templates/exercise/staff/_submissions_table.html#L64
- ??
  https://github.com/apluslms/a-plus/blob/master/course/permissions.py#L143

Here's two alternatives I could think of:

1.
* USE_L10N = True
* Use the same short format and different long formats for English and Finnish
* Define DATETIME_FORMAT, SHORT_DATE_FORMAT etc in aplus/en/formats.py and aplus/fi/formats.py
+Easiest to change
-?

2. 
* USE_L10N = False
* Use the same short and long formats for English and Finnish
* Define DATETIME_FORMAT, SHORT_DATE_FORMAT etc in settings.py
+Consistent
-Date format in English might be confusing


! Disabling L10N does not affect the language of weekday or month.
! When L10N = True, all the formats have to be defined in formats.py files.
! When L10N = False, all the formats have to be defined in settings.py, which creates some issues if we want some formats to be different for different languages.


Best formats in Finnish imo would be:
1. `l j.n.Y H:i`          `keskiviikko 1.7.2020 12:00`	(preferred format in Finnish http://www.kielitoimistonohjepankki.fi/ohje/53)
2. `D j.n.Y H:i`		`ke 1.7.2020 12:00`
3. `r'l d.m.Y \k\l\o H:i'` 	`keskiviikko 1.7.2020 klo 12:00`
4. `l d.m.Y H:i`		`keskiviikko 01.07.2020 12:00`
5. `D d.m.Y H:i`		`ke 01.07.2020 12:00`
6. `j.n.Y l H:i`		`1.7.2020 keskiviikko 12:00` 	(used now)

in English:
1. `l, M j Y, P`		         `Wednesday, Jul 1 2020, noon`
2. `D, M j Y, P`		`Wed, Jul 1 2020, noon` 	(used now)
3. `D, N j, Y, P`		`Wed, July 1, 2020, noon`	(close to django's default format)
4. `l j.n.Y P`			`Wednesday 1.7.2020 noon`
5. `l j.n.Y H:i`		`Wednesday 1.7.2020 12:00` 	(same as in Finnish)

**Comment by ttsirkia** [comment]
I vote number two for Finnish (the current for that is actually bad because of the order) and number two or three for English.

**Comment by Mankro** [comment]
Thanks @veeraalt. I think we should choose your alternative 1 that has `USE_L10N = True` with identical short date formats (1.7.2020) and different long date formats. For the long formats, I prefer the Finnish option 2 (`D j.n.Y H:i`	`ke 1.7.2020 12:00`) and the English option 2 (`D, M j Y, P` `Wed, Jul 1 2020, noon`).

Others could and should share their opinions too, hopefully not any later than today.
@mtthwhggns I hope you would comment and especially on the English format. Is it a problem that the long date format is not the same in all languages since this started from the idea of having consistency across languages?

**Comment by mtthwhggns** [comment]
Thanks @veeraalt and @Mankro . I had planned to reply to this, but was doing some screenreader experiements first:

As far as I'm aware, we're not very tight on space anywhere we use the long date format, so maybe we could avoid the ones with abbreviated versions of weekday and month names, as this will avoid potential problems of screenreaders mispronouncing them ([WCAG 3.1.4](https://www.w3.org/TR/UNDERSTANDING-WCAG20/meaning-located.html)). It's worth noting that different screenreaders perform differently at this task, some read the abbreviated examples as "Wednesday", others just say "Wed". Finnish 2 and 5, and English 1, 2 and 3 use abbreviated forms. 

As I've said a few times in this discussion, I created the issue out of concern for the short date format. Where we do not need the short date format, I prefer the full words, then the format isn't so critical. 

I note that no option in English includes full words - is this intentional?

**Comment by veeraalt** [comment]
Good points @mtthwhggns !

> I note that no option in English includes full words - is this intentional?

Not intentional, I just gathered some examples but it's perfectly fine to use a combination of those, eg. both weekday and month as full words, if using slightly more space is not an issue.

**Comment by raphendyr** [comment]
As a side note, format `r'l d.m.Y \k\l\o H.i'` (`keskiviikko 1.7.2020 klo 12.00`) would be the most correct Finnish to my understanding, however I prefer the colon version (i.e. number 1).

In English, it's typically preferred to use month names instead of numbers, though I personally understand numbers better than the names. Interestingly, `July 1, 2020` (US) feels more natural when read than `1 July 2020` (UK if I understand right). In addition, I think we should use `P` for the time as that creates `am` and  `pm` suffixes. I think `H:i` would be good there too with any kind of date.

**Comment by Mankro** [comment]
Accessibility is important, so let's use full words in long date formats if that doesn't cause other new issues. 

* In Finnish: `keskiviikko 1.7.2020 13:00`
* In English: `Wednesday, July 1, 2020, 13:00`

I don't know if commas are needed in English. Are these formats good?

**Comment by mtthwhggns** [comment]
That would be fine with me. When used in isolation (it’s different inside a sentence), I don’t see a benefit from the commas.

**Comment by ttsirkia** [comment]
In Finnish, there should officially be a dot between hours and minutes, `13.23`, but in my opinion colon is better.

**Comment by raphendyr** [comment]
When implementing this, I think it would be beneficial to edit how format locations are found, so it's easier to edit settings per installation. In addition, we could define defaults in regionally defined folder. Thus, I would move our current definitions to `aplus/region/EU` if we assume these formats would be fine for most of european union, or `aplus/region/FI` if we assume they work only in Finland (note the capitalization). In addition, I would create a configuration to allow local edits. Hence, the settings file should be updated to contain:

```python
FORMAT_MODULE_PATH = [
    'local_region',    # local_region/<language>/formats.py
    'aplus.region.EU', # aplus/region/EU/<language>/formats.py
```

Please comment on the names of the folders as I'm not sure how they should be named. I assume the goal is relatively clear.

**Comment by ttsirkia** [comment]
I don't see the point, which things would be common inside EU? All of these formatting related issues are more or less related to the language, not the continent.

**Comment by Mankro** [comment]
Yeah, no need to over-engineer this. It is enough that formats can be configured for each language and that is the default in Django.

**Comment by raphendyr** [comment]
> All of these formatting related issues are more or less related to the language, not the continent.

We are exactly trying to fix the problem that the formatting is **not** related to langauge, but the region. We need to provide dates in different formats in English in different locations in the world. For example, in installation in Finland, the formats should represent Finnish styles in any language, so they are understandably by Finnish people. I.e. short date format here 01/06/2001 should be read as 1st of June, 2001 and January 6th, 2001 in the US region.  Just for English it could be enough to define locales with the region info (e.g. en_US vs en_GB, en-us and en-gb in django). Though, we actually would need `en_FI`, but that is not a standard. Hence, I believe it's better to not mix language and region info as we can easily use multiple python paths for different sets of configuration.

**Comment by ttsirkia** [comment]
Yes, but the region would not be Europe. For example, for German there would be Germany and Austria. But there is nothing in common between German and Spanish although both languages are spoken inside EU. Therefore, there are `de-DE` and `de-AT` to represent such differences but EU as a region is not anyhow valuable here.

**Comment by raphendyr** [comment]
It could be possible that better solution would be to start supporting language defitions with the region part and then define `en_FI` so we can deviate from `en_GB` for Finnish installations. However, I think that requires to fix how course langauges are mapped to language_country codes (i.e. en -> en_FI and not em -> en_US).

On the region area, the EU might be a bit too huge region as there are some countries that use `dd/mm/yyyy` instead of `dd.mm.yyyy`. I assume that different countries would use a bit different variations of English formats, thus they would require overwriting the en_US/en_GB defitions (e.g. is time presented with 12h or 24h clock).

**Comment by Mankro** [comment]
The discussion around this topic lived a lot. In the end, we decided to change the format files a bit for Finnish and English. The settings.py variable `USE_L10N = True` is kept as before. The formats are not identical in Finnish and English, but we think this is clear and comprehensible. See the PR for details #604. Especially in English we chose to write month names in longer formats so that there is no confusion between the American and British ordering (day-month vs. month-day).
